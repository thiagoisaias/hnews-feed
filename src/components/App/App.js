import React, { Component, Fragment } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Header from "../../components/Header/Header";
import Table from "../../components/Table/Table";
import Spinner from "../../components/Spinner/Spinner";

const Container = styled.div`
  max-width: 900px;
  margin: 0 auto;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  justify-content: center;
`;

const LoadMoreButton = styled.button`
  display: block;
  outline: none;
  border: none;
  cursor: pointer;
  color: #fff;
  background-color: #ff742b;
  border-radius: 3px;
  padding: 6px 12px;
  font-size: 12px;
  margin: 16px auto;
  &:hover {
    opacity: 0.9;
  }
`;

const ErrorMessage = styled.p`
  font-size: 14px;
  color: red;
  text-align: center;
`;

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchTerm: ""
    };
  }

  onSearchChange = event => {
    this.setState({
      searchTerm: event.target.value
    });
  };

  onSearchSubmit = event => {
    event.preventDefault();
    this.props.fetchSearchTopStories(this.state.searchTerm);
  };

  render() {
    const {
      results,
      searchKey,
      error,
      isLoading,
      sortByDate,
      sortByPopularity,
      fetchSearchTopStories
    } = this.props;

    const page =
      (results && results[searchKey] && results[searchKey].page) || 0;
    const list =
      (results && results[searchKey] && results[searchKey].hits) || [];

    if (error) {
      return <ErrorMessage>{error}</ErrorMessage>;
    }

    return (
      <Container>
        <Header
          value={this.state.searchTerm}
          onChange={this.onSearchChange}
          onSubmit={this.onSearchSubmit}
          sortByDate={sortByDate}
          sortByPopularity={sortByPopularity}
        />
        {isLoading ? (
          <Spinner height={54} width={54} />
        ) : (
          <Fragment>
            <Table list={list} page={page} />
            <LoadMoreButton
              onClick={() => {
                fetchSearchTopStories(searchKey, page + 1);
              }}
            >
              Load More
            </LoadMoreButton>
          </Fragment>
        )}
      </Container>
    );
  }
}

App.propTypes = {
  searchKey: PropTypes.string.isRequired,
  results: PropTypes.object.isRequired,
  error: PropTypes.string,
  isLoading: PropTypes.bool,
  sortByDate: PropTypes.func.isRequired,
  sortByPopularity: PropTypes.func.isRequired,
  fetchSearchTopStories: PropTypes.func.isRequired
};

export default App;
