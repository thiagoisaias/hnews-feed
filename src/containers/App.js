import React, { Component } from "react";
import styled from "styled-components";
import Table from "../components/Table";
import Header from "../components/Header";
import FiltersBar from "../components/FiltersBar";
import axios from "axios";

const DEFAULT_QUERY = "React";
const DEFAULT_HPP = "20";
const PATH_BASE = "https://hn.algolia.com/api/v1/search?query=";
const PARAM_PAGE = "page=";
const PARAM_HPP = "hitsPerPage=";

const Container = styled.div`
  max-width: 900px;
  margin: 0 auto;
`;

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      results: null,
      searchKey: "",
      searchTerm: DEFAULT_QUERY,
      error: null
    };
  }

  componentDidMount() {
    const { searchTerm } = this.state;
    this.setState({
      searchKey: searchTerm
    });
    if (this.needsToFetchStories) {
      this.fetchSearchTopStories(searchTerm);
    }
  }

  setSearchTopStories = result => {
    const { hits, page } = result;
    const { searchKey, results } = this.state;

    const oldHits =
      results && results[searchKey] ? results[searchKey].hits : [];
    const updatedHits = [...oldHits, ...hits];

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      }
    });
  };

  needsToFetchStories = searchTerm => {
    console.log(
      "Does it need to fetch new hits?",
      !this.state.results[searchTerm]
    );
    return !this.state.results[searchTerm];
  };

  fetchSearchTopStories = (searchTerm, page = 0) => {
    axios(
      `${PATH_BASE}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`
    )
      .then(response => {
        this.setSearchTopStories(response.data);
      })
      .catch(error => this.setState({ error }));
  };

  onSearchChange = event => {
    this.setState({
      searchTerm: event.target.value
    });
  };

  onSearchSubmit = event => {
    const { searchTerm } = this.state;
    this.setState({
      searchKey: searchTerm
    });
    if (this.needsToFetchStories(searchTerm)) {
      this.fetchSearchTopStories(searchTerm);
    }
    event.preventDefault();
  };

  render() {
    const { results, searchTerm, searchKey, error } = this.state;
    const page =
      (results && results[searchKey] && results[searchKey].page) || 0;
    const list =
      (results && results[searchKey] && results[searchKey].hits) || [];

    if (error) {
      return <p>Something went wrong.</p>;
    }

    return (
      <Container>
        <Header
          value={searchTerm}
          onChange={this.onSearchChange}
          onSubmit={this.onSearchSubmit}
        />
        <FiltersBar />
        <Table list={list} />
        <button onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}>
          Load More
        </button>
      </Container>
    );
  }
}

export default App;
