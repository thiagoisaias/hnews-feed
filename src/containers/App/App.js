import React, { Component, Fragment } from "react";
import styled from "styled-components";
import axios from "axios";
import Header from "../../components/Header/Header";
import Table from "../../components/Table/Table";
import Spinner from "../../components/Spinner/Spinner";

const DEFAULT_QUERY = "React";
const DEFAULT_HPP = "70";
const DEFALT_SEARCH_PARAM = "search?query=";
const PATH_BASE = "https://hn.algolia.com/api/v1/";
const PARAM_PAGE = "page=";
const PARAM_HPP = "hitsPerPage=";

const Container = styled.div`
  max-width: 900px;
  margin: 0 auto;
`;

const LoadMoreButton = styled.button`
  display: block;
  outline: none;
  border: none;
  cursor: pointer;
  color: #fff;
  background-color: #ff742b;
  border-radius: 3px;
  padding: 6px 12px;
  font-size: 12px;
  margin: 16px auto;
  &:hover {
    opacity: 0.9;
  }
`;

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: false,
      results: null,
      searchParam: DEFALT_SEARCH_PARAM,
      searchKey: "",
      searchTerm: DEFAULT_QUERY,
      error: null
    };
  }

  componentDidMount() {
    const { searchParam, searchTerm } = this.state;
    this.setState({
      searchKey: searchTerm
    });
    if (this.needsToFetchStories) {
      this.fetchSearchTopStories(searchParam, searchTerm);
    }
  }

  setSearchTopStories = result => {
    const { hits, page } = result;
    const { searchKey, results } = this.state;

    const oldHits =
      results && results[searchKey] ? results[searchKey].hits : [];
    const updatedHits = [...oldHits, ...hits];

    const filteredUpdatedHits = [...updatedHits].filter(hit => {
      return (
        (hit.title || hit.story_title) &&
        hit.points &&
        hit.author &&
        hit.created_at &&
        (hit.url || hit.story_url)
      );
    });

    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: filteredUpdatedHits, page }
      },
      isLoading: false
    });
  };

  needsToFetchStories = searchTerm => {
    console.log(
      "Does it need to fetch new hits?",
      !this.state.results[searchTerm]
    );
    return !this.state.results[searchTerm];
  };

  fetchSearchTopStories = (searchParam, searchTerm, page = 0) => {
    this.setState({
      isLoading: true
    });
    axios(
      `${PATH_BASE}${searchParam}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`
    )
      .then(response => {
        this.setSearchTopStories(response.data);
      })
      .catch(error => this.setState({ error, isLoading: false }));
  };

  onSearchChange = event => {
    this.setState({
      searchTerm: event.target.value
    });
  };

  onSearchSubmit = event => {
    const { searchParam, searchTerm } = this.state;
    this.setState({
      searchKey: searchTerm
    });
    if (this.needsToFetchStories(searchTerm)) {
      this.fetchSearchTopStories(searchParam, searchTerm);
    }
    event.preventDefault();
  };

  sortByPopularity = () => {
    const { searchKey } = this.state;
    const list = [...this.state.results[searchKey].hits];

    const sortedList = list.sort((a, b) => {
      return b.points - a.points;
    });

    this.setState({
      results: {
        [searchKey]: {
          hits: sortedList
        }
      }
    });
  };

  sortByDate = () => {
    const { searchKey } = this.state;
    const list = [...this.state.results[searchKey].hits];

    const sortedList = list.sort((a, b) => {
      return new Date(b.created_at) - new Date(a.created_at);
    });

    this.setState({
      results: {
        [searchKey]: {
          hits: sortedList
        }
      }
    });
  };

  render() {
    const { results, searchParam, searchTerm, searchKey, error } = this.state;
    const page =
      (results && results[searchKey] && results[searchKey].page) || 0;
    const list =
      (results && results[searchKey] && results[searchKey].hits) || [];

    if (error) {
      return <p>Something went wrong.</p>;
    }

    return (
      <Container>
        <Header
          value={searchTerm}
          onChange={this.onSearchChange}
          onSubmit={this.onSearchSubmit}
          sortByDate={this.sortByDate}
          sortByPopularity={this.sortByPopularity}
        />
        {this.state.isLoading ? (
          <Spinner height={48} width={48} />
        ) : (
          <Fragment>
            <Table list={list} />
            <LoadMoreButton
              onClick={() =>
                this.fetchSearchTopStories(searchParam, searchKey, page + 1)
              }
            >
              Load More
            </LoadMoreButton>
          </Fragment>
        )}
      </Container>
    );
  }
}

export default App;
