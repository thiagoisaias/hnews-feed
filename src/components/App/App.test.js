import React from "react";
import Enzyme, { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import App from "./App";

describe("App", () => {
  it("renders properly", () => {
    const props = {
      searchKey: "Hello",
      results: {},
      error: null,
      isLoading: false,
      sortByDate: jest.fn(),
      sortByPopularity: jest.fn(),
      setSearchKey: jest.fn(),
      fetchSearchTopStories: jest.fn()
    };
    const component = shallow(<App {...props} />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });
});
