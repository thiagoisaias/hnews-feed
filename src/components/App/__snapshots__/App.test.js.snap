// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders properly 1`] = `
<styled.div>
  <Header
    onChange={[Function]}
    onSubmit={[Function]}
    sortByDate={[Function]}
    sortByPopularity={[Function]}
    value=""
  />
  <React.Fragment>
    <Table
      list={Array []}
      page={0}
    />
    <styled.button
      onClick={[Function]}
    >
      Load More
    </styled.button>
  </React.Fragment>
</styled.div>
`;
